### PR TITLE
[Rendy] Change DynamicVertex to a generic type for index buffers

### PR DESCRIPTION
**Pull request by jaynus**
## Description

The current implementation of `DynamicBuffer` is hard-coded to only create vertex buffers, and has some offsets hard coded. This implementation makes this instead generic `DynamicVertexData` (and `DynamicVertexBuffer` / `DynamicIndexBuffer`), which statically dispatches to the appropriate `bind` function of the rendy encoder based on its type parameter. This lets us dispatch for the correct usage at compile time without any hotpath overhead. 

## Additions

- Adds the generic parameters and other associated data in submodules::vertex

## Modifications

- Changes the struct tuple inner of amethyst_rendy::submodules::texture::TextureSub to public.

## PR Checklist

By placing an x in the boxes I certify that I have:

- [x] Ran `cargo test --all` locally if this modified any rs files.
- [x] Ran `cargo +stable fmt --all` locally if this modified any rs files.
- [x] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.

### CODE REVIEW

File: amethyst_rendy/src/submodules/texture.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ enum TextureState<B: Backend> {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
-pub struct TextureId(u32);
+pub struct TextureId(pub u32);
```
Comment (Frizi) [inline amethyst_rendy/src/submodules/texture.rs]: This wrapper type is done specifically to not allow constructing arbitrary texture ids out of thin air. The inner type should not be public

File: amethyst_rendy/src/submodules/vertex.rs [hunk 2]
```diff
@@ -11,25 +14,121 @@ use crate::{
...
+    }
+}
+
+/// This structure wraps [PerImageDynamicVertexData], managing multiple instances and providing
+/// an easy-to-use interface for having per-image buffers. This is needed because multiple images
+/// (frames) can be in flight at any given time, so multiple buffers are needed for the same data.
 #[derive(Debug)]
-struct PerImageDynamicVertex<B: Backend> {
-    buffer: Option<Escape<Buffer<B>>>,
+pub struct DynamicVertexData<B: Backend, V: VertexDataBufferType<B>, T: 'static> {
```
Comment (Frizi) [inline amethyst_rendy/src/submodules/vertex.rs]: i believe you don't need `T` for anything here.

File: amethyst_rendy/src/submodules/vertex.rs [hunk 3]
```diff
@@ -70,35 +169,53 @@ impl<B: Backend, T: 'static> DynamicVertex<B, T> {
...
         &self,
         index: usize,
         binding_id: u32,
+        offset: u64,
         encoder: &mut RenderPassEncoder<'_, B>,
     ) -> bool {
         self.per_image
             .get(index)
-            .map_or(false, |i| i.bind(binding_id, encoder))
+            .map_or(false, |i| V::bind(binding_id, encoder, &i.buffer, offset))
```
Comment (Frizi) [inline amethyst_rendy/src/submodules/vertex.rs]: it feels wrong to me that `binding_id` is being used only sometimes here. I think the whole `bind` method should be split in half and implemented only for relevant types. One for each impl of
```rs
impl<B: Backend, T: 'static> DynamicVertexData<B, VertexData<T>> { .. }
impl<B: Backend, T: 'static> DynamicVertexData<B, IndexData<T>> { .. }
```

File: amethyst_rendy/src/submodules/texture.rs (hunk 1 above)
Comment (jaynus) [reply amethyst_rendy/src/submodules/texture.rs]: Removed.

File: amethyst_rendy/src/submodules/vertex.rs (hunk 2 above)
Comment (jaynus) [reply amethyst_rendy/src/submodules/vertex.rs]: It's used below at `I::Item: AsRef<[T]>,`

File: amethyst_rendy/src/submodules/vertex.rs (hunk 3 above)
Comment (jaynus) [reply amethyst_rendy/src/submodules/vertex.rs]: I've changed up the implementation now to specialize for each case and call the appropriate binding function, so there is no spare random value now.

File: amethyst_rendy/src/submodules/vertex.rs [hunk 4]
```diff
@@ -69,36 +157,76 @@ impl<B: Backend, T: 'static> DynamicVertex<B, T> {
...
+        })
     }
 }
 
-impl<B: Backend> PerImageDynamicVertex<B> {
+/// an easy-to-use interface for managing, growing and binding a given vertex buffer type. This
+/// implementation also leverages the [VertexDataBufferType] trait type for statically dispatching
+/// the appropriate binding and allocation functions, preventing hot-path branching.
+#[derive(Debug)]
+pub struct PerImageDynamicVertexData<B: Backend, V: VertexDataBufferType> {
```
Comment (Frizi) [inline amethyst_rendy/src/submodules/vertex.rs]: this should not be `pub`

**Comment by jaynus** [comment]
bors r=jojolepro,Frizi

**Pull request merged**